core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -589,10 +594,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
         }
 
         // generate id if not already provided and id generation is allowed
-        if (allowIdGeneration) {
-            if (id == null) {
-                id(UUIDs.base64UUID());
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedID == false;
+            autoGeneratedID = true;
```

Comment:
I dont' think we should do this to be honest I added assertions along the way to ensure it's only true if `autoGeneratedID` is true too also it's documented. We can switch so something like:

```
enum IDSource {
  External, AutoGenerated, RetryAutoGenerated
}
`` 
to have the different states clear, WDYT?